java/code/src/com/redhat/rhn/manager/distupgrade/DistUpgradeManager.java

```diff
@@ -695,4 +697,42 @@ public static Long scheduleDistUpgrade(User user, Server server,
         // Return the ID of the scheduled action
         return ActionManager.scheduleDistUpgrade(user, server, details, earliest).getId();
     }
+
+    /**
+     * Remove incompatible migration targets compared to the installed products.
+     * Write the failed products in the missingSuccesorExtensions set in case it should
+     * be shown somewhere.
+     *
+     * @param installedProducts Optional set of the installed products
+     * @param allMigrationTargets all calculated migration targets
+     * @param missingSuccessorExtensions OUT: info about installed extensions missing a successor
+     * @return list of valid migration targets
+     */
+    public static List<SUSEProductSet> removeIncompatibleTargets(Optional<SUSEProductSet> installedProducts,
```

Comment:
You can pass the `installedProducts` directly, not via an optional (the only thing that's done via the optional is `get()` in this method).

Reply:
Right, where we call it we have it as optional value. Somebody refactored in the past the code to have installedProducts as Optional. So I keep compatibility.